[PEP 747] Recognize TypeForm[T] type and values (#9773)

davidfstr opened the pull request:
> 👉 **Superceded**: This PR is replaced by #19596 👈 

Implements the [TypeForm PEP 747](https://peps.python.org/pep-0747/), as an opt-in feature enabled by the CLI flag `--enable-incomplete-feature=TypeForm`.

Implementation approach:

* The `TypeForm[T]` is represented as a type using the existing `TypeType` class, with an `is_type_form=True` constructor parameter. `Type[C]` continues to be represented using `TypeType`, but with `is_type_form=False` (the default).

* Recognizing a type expression literal such as `int | str` requires parsing an `Expression` as a type expression. Only the SemanticAnalyzer pass appears to have the ability to parse type expressions, using `SemanticAnalyzer.expr_to_analyzed_type()`. In particular the later `TypeChecker` pass does not appear to have the ability to parse type expressions.

* Therefore during the SemanticAnalyzer pass, at certain syntactic locations (i.e. assignment r-values, callable arguments, returned expressions), the analyzer tries to parse the `Expression` it is looking at using `try_parse_as_type_expression()` - a new function - and stores the result (a `Type`) in `Expression.as_type` - a new attribute.

* During the later TypeChecker pass, when looking at an `Expression` to determine its type, if the expression is in a type context that expects some kind of `TypeForm[...]` and the expression was successfully parsed as a type expression by the earlier SemanticAnalyzer pass, the expression will be given the type `TypeForm[expr.as_type]` rather than using the regular type inference rules for a value expression.

* Key relationships between `TypeForm[T]`, `Type[C]`, and `object` types are defined in the visitors powering `is_subtype`, `join_types`, and `meet_types`.

* The `TypeForm(T)` expression is recognized as a `TypeFormExpr` and has the return type `TypeForm[T]`.

* The new test suite in `check-typeform.test` is a good reference to the expected behaviors for operations that interact with `TypeForm` in some way.

Controversial parts of this PR, in @davidfstr 's opinion:

* Type form literals are only recognized in certain syntactic locations (and not ALL possible locations). Namely they are recognized as (1) assignment r-values, (2) callable expression arguments, and (3) as returned expressions, but nowhere else. For example they aren't recognized in expressions like `dict_with_typx_keys[int | str]`.

* SemanticAnalyzer makes very many unnecessary calls to try_parse_as_type_expression() - a call for every syntactic location where recognizing a type form literal is supported, potentially doubling the type to parse value expressions. It would be much preferred if `try_parse_as_type_expression()` could only be called when TypeChecker is looking at an expression known to be in a `TypeForm[T]` type context. Unfortunately TypeChecker cannot parse type expressions itself; only the earlier SemanticAnalyzer pass can do that.

* `try_parse_as_type_expression()` saves and restores a lot of SemanticAnalyzer state, so that failing to parse a type expression literal doesn't cause unexpected side effects.
    - The way that the save/restore of state happens makes me wonder if incremental typechecking would correctly detect small changes made to a type expression literal parsed by `try_parse_as_type_expression()`.

* ~~The `Expression` class no longer declares (empty) `__slots__` and now always includes an `as_type` attribute~~ **(Edit: Fixed.)**
    - ~~I wonder if this will increase memory usage for `Expression` objects enough to noticeably decrease mypy performance. It's unclear to me what tools are available to check for such an issue.~~

* The existing `TypeType` class is now used to represent BOTH the `Type[T]` and `TypeForm[T]` types, rather than introducing a distinct subclass of `Type` to represent the `TypeForm[T]` type. This was done to simplify logic that manipulates both `Type[T]` and `TypeForm[T]` values, since they are both manipulated in very similar ways.

* The "normalized" form of `TypeForm[X | Y]` - as returned by `TypeType.make_normalized()` - is just `TypeForm[X | Y]` rather than `TypeForm[X] | TypeForm[Y]`, differing from the normalization behavior of `Type[X | Y]`.

davidfstr committed:
[PEP 747] Recognize TypeForm[T] type and values (#9773)

User must opt-in to use TypeForm with --enable-incomplete-feature=TypeForm

In particular:
* Recognize TypeForm[T] as a kind of type that can be used in a type expression
* Recognize a type expression literal as a TypeForm value in:
    - assignments
    - function calls
    - return statements
* Define the following relationships between TypeForm values:
    - is_subtype
    - join_types
    - meet_types
* Recognize the TypeForm(...) expression
* Alter isinstance(typx, type) to narrow TypeForm[T] to Type[T]

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

davidfstr commented:
There are many unexpected test suite errors when run in CI. I'll take a look in the next few days.

davidfstr committed:
Eliminate use of Type Parameter Syntax, which only works on Python >= 3.12

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

davidfstr committed:
Remove INPUTS directory

JukkaL commented:
> Unfortunately TypeChecker cannot parse type expressions itself; only the earlier SemanticAnalyzer pass can do that.

What are the reasons for this? Maybe we can adjust something to make it possible to analyze type expressions in TypeChecker.

> Type form literals are only recognized in certain syntactic locations (and not ALL possible locations).

Is this specified in the PEP? If they should be allowed in other contexts, we should try to make it possible. For example, if we could parse type expressions in TypeChecker, we could perhaps use the type context to decide whether something might be a TypeForm. The idea would be like this (haven't thought about this very carefully though):
* If the type context is `TypeForm[...]` (or some type context union item is `TypeForm[...]`), attempt to process the expression as a type form. If successful, infer `TypeForm[...]` as the type of the expression. Otherwise only try to analyze the expression as a normal expression (similar to the old behavior).
* I guess processing an expression as a type form during type checking may require duplicating a bunch of functionality in `mypy.typeanal`, but not everything there seems necessary. Do you think it's feasible to implement a new visitor that converts semantically analyzed expressions to types? Anyway even if we have to reimplement most of `mypy.typeanal`, it may be better overall compared to doing a lot of redundant work during semantic analysis. Since we can likely reuse some of the work done by semantic analyzer (name binding at least), this should be easier than what `mypy.typeanal` currently does.

> The Expression class no longer declares (empty) `__slots__` and now always includes an `as_type` attribute

This will increase memory use (on average 8 bytes per expression node in compiled mypy). Again, I'm in favor of finding a way to avoid this.

JukkaL commented:
And thanks for working on this! This will be a useful feature, and will enable many new use cases.

davidfstr committed:
WIP: Don't declare attributes on Expression to avoid confusing mypyc

davidfstr committed:
SQ -> WIP: Don't declare attributes on Expression to avoid confusing mypyc

Recognize the remaining subtypes of MaybeTypeExpression.

davidfstr committed:
SQ -> WIP: Don't declare attributes on Expression to avoid confusing mypyc

Fix: Workaround mypy thinking incorrectly that isinstance(X, Union[...])
does not work at runtime.

davidfstr commented:
> > Unfortunately TypeChecker cannot parse type expressions itself; only the earlier SemanticAnalyzer pass can do that.
> 
> What are the reasons for this? Maybe we can adjust something to make it possible to analyze type expressions in TypeChecker.

Analyzing an Expression as a type expression currently requires both `expr_to_unanalyzed_type()` to get an "unanalyzed type" and `SemanticAnalyzer.anal_type()` to make that into an "analyzed type", which is what we want.

The logic of `anal_type()` inside SemanticAnalyzer looks pretty complicated. It has a `TypeAnalyser` object inside it which has a back-reference to the enclosing SemanticAnalyzer object (as a SemanticAnalyzerCoreInterface). It's unclear to me whether TypeAnalyser could be decoupled from SemanticAnalyzer to allow it to be used during the TypeChecker pass.

I agree that it would be ideal if TypeAnalyser could be used directly inside TypeChecker. It would eliminate several of the controversial issues mentioned at the top of this PR.

> > Type form literals are only recognized in certain syntactic locations (and not ALL possible locations).
> 
> [...] we should try to make it possible. For example, if we could parse type expressions in TypeChecker, we could perhaps use the type context to decide whether something might be a TypeForm.

Agreed.

The PEP doesn't limit the syntactic locations where type form literals should be recognized, so the current implementation which only recognizes in certain locations isn't ideal.

> > The Expression class no longer declares (empty) __slots__ and now always includes an as_type attribute
> 
> This will increase memory use (on average 8 bytes per expression node in compiled mypy).

The latest version of this PR avoids the issue by moving the "as_type" attribute out of `Expression` and to the 4 specific subtypes that could be the root of a type expression (i.e. `IndexExpr, NameExpr, OpExpr, StrExpr`).

Again, if TypeChecker could parse type expression itself, there would be no need to store a parsed `Type` on an `Expression` (subclass) at all.

JukkaL commented:
> Again, if TypeChecker could parse a type expression itself there would be no need to store a parsed Type on an Expression (subclass) at all.

I'm no longer sure if this is easy to do, since string literal types wouldn't be semantically analyzed, and so we'd need to parse and analyze them from scratch during type checking, and doing this could be tricky. Your current approach may be better after all.

To make my alternative approach work with string literal types would imply calling the semantic analyzer from the type checker to analyze AST subtrees that are used in a type form context. I'm not sure if this is practical to do, but it might be. If we can do this, then we could probably use the existing type analyzer as well, so the implementation would be simpler. This could be the best approach (if practical):
 1. It would be not require writing another type analyzer that works on semantically analyzed ASTs.
 2. We wouldn't have to perform any extra work during semantic analysis.
 
 I'll think about this more.

davidfstr commented:
I investigated whether a TypeAnalyzer could be created inside the TypeChecker pass. It looks like it _might_ be possible:

* `TypeAnalyzer.__init__` requires notably an `api: SemanticAnalyzerCoreInterface` and a `tvar_scope: TypeVarLikeScope`.
* `SemanticAnalyzerCoreInterface` I can mostly implement in a TypeChecker context. The tricky abstract methods are:
    * `.lookup_qualified()` - SemanticAnalyzer does a lookup inside {global_decls, nonlocal_decls, locals, globals}. The similar TypeChecker.lookup_qualified() does a lookup inside {modules, globals}.
    * `.lookup_fully_qualified_or_none()` - SemanticAnalyzer does a lookup inside {modules}. The similar TypeChecker.lookup_qualified() does a lookup inside {modules, globals}.
    * `.type()` - SemanticAnalyzer tracks the type of the innermost class scope in this property. TypeChecker could probably be modified to do similar bookkeeping.
* I'm not sure how to get a `TypeVarLikeScope` in a `TypeChecker` context.

It's not clear to me that the behavior of SemanticAnalyzer's lookup methods are aligned enough with TypeChecker.lookup_qualified() to make it reasonable to implement the former on top of the latter. @JukkaL, any insights on this question?

davidfstr committed:
Recognize non-string type expressions everywhere lazily. Optimize eager parsing to bail earlier.

Also:
* No longer add "as_type" attribute to NameExpr and MemberExpr.
  Retain that attribute on StrExpr, IndexExpr, and OpExpr.
* Recognize dotted type expressions like "typing.List".

JukkaL commented:
> I investigated whether a TypeAnalyzer could be created inside the TypeChecker pass.

This may be impractical, since symbol tables for local variables (including type aliases) are only available during semantic analysis -- and we don't want to keep them around after semantic analysis. I don't see a good way of resolving string literal types that refer to local type aliases during type checking. So your original idea together with my suggestions above may still be the best way forward. This means that `lookup_qualified` probably can't be made to work with locals inside TypeChecker.

Code like this seems tricky unless string literal types are resolved during semantic analysis:
```py
def f() -> None:
    A: TypeAlias = int | str
    foo('list[A]')  # foo() expects a type form
```

davidfstr committed:
NOMERGE: Disable test that is already failing on master branch

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

davidfstr committed:
Fix mypyc errors: Replace EllipsisType with NotParsed type

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

davidfstr committed:
mypy_primer: Enable TypeForm feature when checking effects on open source code

davidfstr commented:
@JukkaL , ready for another review pass. I've made the following changes:

* The SemanticAnalyzer pass continues to greedily/speculatively parse expressions as type expressions in certain syntactic locations (assignment r-values, call arguments, returned expressions). However it has new heuristics for quickly skipping common expression patterns that could not possibly be a valid type expression. It also skips parsing NameExpr and MemberExpr because those can always be parsed later in TypeChecker...
* The TypeChecker pass (in ExpressionChecker) now lazily parses expressions as type expressions when in a TypeForm type context. This later parsing code can recognize any kind of type expression EXCEPT those containing a string annotation. If it encounters a potential type expression that wasn't already parsed by the SemanticAnalyzer pass, a note is emitted (`TypeForm containing a string annotation cannot be recognized here. Try assigning the TypeForm to a variable and use the variable here instead.`).

In summary:
* Type expressions without string annotations can now be lazily (and efficiently) recognized everywhere.
* Type expressions with string annotations can only be recognized (greedily) in (1) assignment r-values, (2) call arguments, and (3) returned expressions.
* There is overhead for greedily trying to parse every expression in the above syntactic locations as a type expression, but I've tried to minimize that overhead with heuristics that bail out of parsing quickly for common expression patterns.
* Memory use is slightly increased in the AST because {StrExpr, IndexExpr, OpExpr} gain 1 additional `as_type` attribute slot.

davidfstr committed:
Revert "mypy_primer: Enable TypeForm feature when checking effects on open source code"

This reverts commit 41342501be534e67eb895c5390d4fac9bf54630c.

davidfstr committed:
NOMERGE: mypy_primer: Enable --enable-incomplete-feature=TypeForm when checking open source code

davidfstr committed:
Ignore warnings like: <type_comment>:1: SyntaxWarning: invalid escape sequence '\('

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

leycec commented:
@davidfstr: @beartype is now down to only **31 air-quoted "errors"** against this PR. Hallelujah! We give praise to Guido. That said, more than a few of these "errors" appear to be false positives in this PR's current implementation of `typing.TypeForm`. Notably, this PR appears to be:

* Failing to match `typing.Any` as a valid `typing.TypeForm`. `Any` had better be a valid type hint. Right? I will eat my own mustache if `Any` isn't a valid type hint. Everywhere @beartype attempts in vain to return `Any` from a function annotated as returning `TypeForm[Any]`, this PR loudly complains that:

  ```
  beartype/_check/convert/_convcoerce.py:191: error: Incompatible return value type
  (got "<typing special form>", expected "TypeForm[Any]")  [return-value]
  ```

I'll report back if I find any other stragglers. Altogether and all in all, though, I'm *exceptionally* impressed with this madness. Thanks so much for all your valiant efforts here! Together, we'll build a better type-checker. And it shall be known as... `mybeartypepy.` :smile:

davidfstr commented:
> The PEP 484-compliant `typing.Any` singleton as a valid `typing.TypeForm`.

> Any PEP 593-compliant `typing.Annotated[...]` type hint as a valid `typing.TypeForm`.

Indeed both of these should be recognized as TypeForms. I'll investigate.

> Any **stringified type hint** as a valid `typing.TypeForm`.

Very interesting. Stringified type hints are already pretty heavily tested since they're difficult to support, so beartype must be using them in a way I did not anticipate. I'll investigate.

> Lastly, a super-minor one: the `Union.__getitem__()` dunder method could profitably be annotated as returning a `TypeForm` of some sort.

Indeed. There are a lot of things in typeshed that will be possible to annotate more-precisely as `TypeForm` soon. And I've already seen some positive support for introducing such annotations in typeshed after this PEP is (hopefully) officially accepted.

@leycec Thanks for investigating/fixing the other mypy errors in beartype! I'll get to work looking at the remaining air-quoted "errors".

davidfstr committed:
Rerun CI

davidfstr commented:
* To fix the remaining mypy_primer errors in beartype **I believe typing_extensions needs to release a version to PyPI that includes TypeForm (>= 2024-09-28)**.
* I tried to alter mypy_primer to forcefully install a later version of typing_extensions with beartype but couldn't figure out the appropriate incantation in `mypy_primer/projects.py`.

In the meantime, ready for another review pass. @JukkaL ? Or is there another mypy maintainer I should invite to take a look?

leycec commented:
> To fix the remaining `mypy_primer` errors in beartype I believe `typing_extensions` needs to release a version to PyPI that includes `TypeForm` (>= 2024-09-28).

**Truth.** At python/typing_extensions#518, I publicly (and pathetically) begged `typing-extensions` maintainers to release a new version. That was over three months ago. They... *did nothing*. It's been nearly a year since their last release. Something has gone fundamentally wrong. They're clearly in [DevHell™](https://en.wikipedia.org/wiki/Development_hell). What's unclear is how they can extricate themselves. :face_exhaling: 

At least `mypy` and @beartype are doing the right thing. We all still regularly release updates. Everybody here is on the ball. Thanks again to @davidfstr for championing this essential integration with runtime type-checkers. You're awesome, Dave.

davidfstr committed:
Improve warning message when string annotation used in TypeForm context

davidfstr commented on test-data/unit/check-typeform.test:
```diff
@@ -218,19 +218,32 @@ dict_with_typx_keys[int | str] += 1
+from typing import Any, Dict, List, TypeForm
+types: Dict[str, TypeForm] = {'any': Any}
+# Ensure warning can be ignored if does not apply.
+list_of_typx1: List[TypeForm] = [types['any']]  # N: TypeForm containing a string annotation cannot be recognized here. Surround with TypeForm(...) to recognize.
+list_of_typx2: List[TypeForm] = [types['any']]  # type: ignore[maybe-unrecognized-str-typeform]
```
I notice that mypy doesn't actually say what the code is for this note, so it wouldn't be obvious to a user how to suppress it. Presumably mypy doesn't emit codes for notes in general (as opposed to errors).

```
INPUTS/test_typeform.py:7: note: TypeForm containing a string annotation cannot be recognized here. Surround with TypeForm(...) to recognize. 👈 NO ERROR CODE
INPUTS/test_typeform.py:7: error: List item 0 has incompatible type "str"; expected "TypeForm[Any]"  [list-item]
```

**I'll look into altering mypy to emit codes for notes too.**

davidfstr committed:
Print error code for the MAYBE_UNRECOGNIZED_STR_TYPEFORM note

davidfstr committed:
Document the [maybe-unrecognized-str-typeform] error code

davidfstr committed:
Fix doc generation warning

davidfstr commented:
This PR to implement TypeForm in mypy is still waiting on review after ~3 weeks.

@JukkaL did you want to take a look, since I already incorporated some feedback from you?

davidfstr commented:
@ilevkivskyi or @hauntsaninja would either of you be available to review this PR? (Both of you have contributed to this repository recently so I'm assuming you have write access.) Ideally I'd like JukkaL's input but I haven't been able to reach him in a month.

ilevkivskyi commented:
Sorry for delay. We are currently busy with finalizing the 1.16 release. I guess either me or @JukkaL should be able to review this once we are done with 1.16.

davidfstr committed:
Merge branch 'master' into f/typeform3

davidfstr commented:
Update: I got feedback from Jukka in-person at PyCon US around 5/16. I'm planning to integrate that feedback to this PR over the next few weeks.

I may opt to open an entirely new PR with the new revisions, since the original implementation has changed a lot from the feedback I've received already.

leycec commented:
Thanks so much for all your tremendous work here and elsewhere, @davidfstr. I hope you had an awesome time at PyCon, too. I'd love to be able to meet up with everyone there someday, but... I'm kinda introverted. This shell is thick. :sweat_smile: 

**Back to PEP 747.** Sadly, [mypy began publicly breaking PEP 747 support in @beartype four days ago](https://github.com/beartype/beartype/actions/runs/15340418592). Oddly, this *only* appears to be happening remotely under GitHub Actions-hosted continuous integration (CI) workflows. I can't replicate this breakage locally – even after updating `mypy` and `typing-extensions` to their most recent stable releases. `mypy 1.16.0` runs fine locally. `mypy 1.16.0` hates @beartype remotely. Since I'm at a bit of a loss as to what to do, I've **temporarily disabled mypy under remote CI.**

This fills me with deep regret. I've come to love and adore mypy. <sup>...unlike `pyright`, which mostly just wastes my time with false positives. `</ahem>`</sup> Mypy keeps @beartype honest. Sadly, @beartype no longer knows how to satisfy mypy's unbidden desires.

If anyone has any insight on what @beartype can do to make mypy feel better, drop me a line either here or at the [@beartype issue tracker](https://github.com/beartype/beartype/issues). Until then, thanks again for all your tremendous volunteerism, @davidfstr! You da QA MVP. :smile:

davidfstr commented:
This PR is superceded by #19596
